### ISSUE TITLE
Does the parameter num_classes in simclr config has any significance in actual clustering?

### ISSUE THREAD

**Issue by isldev** [wvangansbeke/Unsupervised-Classification#25]
Hi,
I am ML student, trying to sort a custom unlabelled dataset using this technique.I noticed in the pretext method (using **simclr** in my case) we have the parameter `num_classes` in the config file, as I understand the actual clusters are made on the second step, so I wanted to ask if this parameter has any high significance on the overall clustering? Also just to be sure, the accuracy metrics on **simclr** is based on the ground truth, yes?
Thank you for the papers and the codes. its really cool :)

edit: grammar error

**Comment by silvia1993** [comment]
Hi @wvangansbeke,

I have a question about the evaluation performed after each epoch in the first step when the weighted_knn method is called.

What exactly happens when I call this method? Is only the accuracy of the pretext task?
I get ~ 99% from this call, but, in the end, I have:
"The accuracy of the first 20 closest neighbors on the train is 20.22"

Thank you!

**Comment by wvangansbeke** [comment]
Hi @silvia1993,

- The [weighted KNN metric](https://github.com/wvangansbeke/Unsupervised-Classification/blob/master/utils/memory.py#L21) is computed for the validation set and looks for the nearest neighbors in the training set. In fact, it will look for the K nearest samples (K = 100 in this repo) based on feature similarity (=cosine similarity). Then it weighs the labels of the K nearest training samples with their corresponding distances. So, to compute this metric, you need all the labels of the training set. It basically checks how many of nearest neighbors in the training set have the same label as a particular test sample and it weighs them exponentially with their distances.

- The [nearest neighbors accurcay](https://github.com/wvangansbeke/Unsupervised-Classification/blob/master/utils/memory.py#L46) is only computed on the train set. It does not take into account distances and simply checks how many neighbors have the same label as its anchor. 

Normally both metrics should increase during training. It's hard to say if it's the case for your particular problem though. Maybe have a look at cifar10 first and check how these metrics ought to behave during training. 

Indeed, 20% is low while 99% is suspiciously high. You can see if reducing the amount of neighbors provides a lot of improvement (which should be the case. If not, there is probably something wrong). You can probably do some sanity checks too. Also, definitely try to visualize a few nearest neighbors on your own dataset after training to check if the quantitative results make sense. 

Good luck!
